### PR TITLE
Geom: Port `Polynomial`

### PR DESCRIPTION
**Pull request by jsn0716**
### Description
Port over `Polynomial` from #942 with modifications:

- `Polynomial` class is now more flexible. If a user creates a `Polynomial` representing `x^2 + 1`, they can set the coefficient of `x^3` term after the creation of the `Polynomial`, or the coefficient of any term.
  - Despite the discussion on Saturday about the ordering of coefficients, to accommodate this flexibility, I chose to have the coefficients in the ascending order of the term degrees. If it was stored in the descending order, every time a user sets the coefficient of a term outside the range of the initial `Polynomial` construction would have resulted in shifting the entire `vector`. 
  - This also removes the restriction that the first coefficient in constructing `Polynomial` has to not be 0
- The `Polynomial` class is now accessed with `getCoeff` and `setCoeff`, so the user does not deal with a coefficient `vector`. The only time the user would deal with a `vector` is for calling the `constructor`.
- Now supports the basic operations `+`, `-`, `*`, `+=`, `-=`, `*=`, `==`
- Does not yet support: `differentiate`, `integrate`, `division`, transformations like `scale` and `translate`
- Default constructor creates a zero polynomial, `P(x) = 0`

### Testing Done
Tests for every function in `polynomial_test.cpp`

### Resolved Issues
Part 2/n of #957

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/new_geom/polynomial.h
Comment (jsn0716) [inline src/software/new_geom/polynomial.h]: Will be changed once #970 is merged, where `EPSILON` is defined in `GeomConstants`

File: src/software/new_geom/polynomial.cpp
Comment (jonathanlew) [inline src/software/new_geom/polynomial.cpp:51]: Please do that!

File: src/software/new_geom/polynomial.h
Comment (jonathanlew) [inline src/software/new_geom/polynomial.h]: what if this isn't true? are we pruning?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Why does this need to be in new geom? Does it depend on new_geom specific things? Since splines are the only thing dependent on polynomials, I'd prefer integrating it right away if possible. My understanding is that Point/Vector is staged in new_geom because it would conflict with existing point vector and integrating immediately would be hard. @MathewMacDougall @garethellis0 Thoughts?

>       * This also removes the restriction that the first coefficient in constructing `Polynomial` has to not be 0
why? doesn't look like we're pruning coeffs here

File: src/software/new_geom/polynomial.h
Comment (jsn0716) [reply src/software/new_geom/polynomial.h]: Forgot to remove that comment, so here the `size` would not equal the degree of the polynomial, you would need to use `getDegree`

**Comment by jsn0716** [comment]
@jonathanlew the `Line` class is going to be a subclass of `Polynomial`
And I didn't bother pruning but it doesn't make it an invalid `Polynomial`. If a user creates a `Polynomial` with `{3, 2, 1, 0}` equivalent to `3 + 2x + x^2 + 0`, `getDegree` still finds the right degree. I can also add back pruning if needed.

File: src/software/new_geom/polynomial.cpp
Comment (jonathanlew) [inline src/software/new_geom/polynomial.cpp]: I see, you check order here. If store this outside the loop (so we don't call this every loop), then I think it's OK.

**Review by jonathanlew — CHANGES_REQUESTED** [review]

File: src/software/new_geom/polynomial.cpp
Comment (jsn0716) [reply src/software/new_geom/polynomial.cpp]: Addressed in 37e8554cd0bbfba2772160f2707effbccccc

File: src/software/new_geom/polynomial.h
Comment (jsn0716) [reply src/software/new_geom/polynomial.h]: Addressed in 37e8554cd0bbfba2772160f2707effbccccc

File: src/software/new_geom/polynomial.cpp
Comment (jsn0716) [reply src/software/new_geom/polynomial.cpp:51]: Done: 37e8554cd0bbfba2772160f2707effbccccc

**Comment by garethellis0** [comment]
> Why does this need to be in new geom? Does it depend on new_geom specific things? Since splines are the only thing dependent on polynomials, I'd prefer integrating it right away if possible. My understanding is that Point/Vector is staged in new_geom because it would conflict with existing point vector and integrating immediately would be hard. @MathewMacDougall @garethellis0 Thoughts?

@jonathanlew - good points. The idea here was that this was going to be integrated into the new geom hierarchy, which means that it must depend on the new `Point/Vector` implementations. With that in mind, if we were to use it in the old geom library, this would mean that anything using it would be using two different implementations of `Point`/`Vector`. Basically, I think that if we can cut over all the usages of `Point` and `Vector` to the new implementations, we can integrate this class into the old geom stuff immediately. @jsn0716 @scveloso @MathewMacDougall - your thoughts please?

File: src/software/new_geom/BUILD
Comment (garethellis0) [inline src/software/new_geom/BUILD:18]: We shouldn't have to add this, everything in the codebase should already be built with it. Does removing this cause an error?

File: src/software/new_geom/polynomial.h
Comment (garethellis0) [inline src/software/new_geom/polynomial.h]: I commented on this over in the `Spline` PR, but I still disagree with having a function interface where it's trivial for the user to pass in entirely invalid values (ex. `{0,1}, {0,2}`). I'm also not sure that there's really a need for a special case method for 1st order polynomials. Thoughts on just having the user call the constructor with coefficients? (@jonathanlew - FYI since we were discussing this over on the `Spline` PR).

File: src/software/new_geom/polynomial.h
Comment (garethellis0) [inline src/software/new_geom/polynomial.h]:
``` suggestion
     * @return the index-1 of the highest order non-zero coefficient
```

File: src/software/new_geom/polynomial.cpp
Comment (garethellis0) [inline src/software/new_geom/polynomial.cpp]: nit; just did a bit of a double-take while reading this, could we change to:
``` suggestion
    return (order >= coeffs.size()) ? 0 : coeffs[order];
```

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good, nothing major from my end.

File: src/software/new_geom/polynomial.cpp
Comment (jonathanlew) [reply src/software/new_geom/polynomial.cpp]: Please don't use tertiary operators. Clarity > line count

File: src/software/new_geom/polynomial.h
Comment (jonathanlew) [reply src/software/new_geom/polynomial.h]: This was made for convenience when constructing splines from points. We should have a polynomial spline constructor that interpolates points with given degree polynomials and uses right number continuity and arbitrary boundary conditions to determine coefficients. Then we can use the coefficient constructor

Comment (jonathanlew) [reply src/software/new_geom/polynomial.h]: If this stays in new_geom until we add the function I described, then this constructor can be removed

**Review by jonathanlew — CHANGES_REQUESTED** [review]

**Comment by jsn0716** [comment]
> @jonathanlew - good points. The idea here was that this was going to be integrated into the new geom hierarchy, which means that it must depend on the new `Point/Vector` implementations. With that in mind, if we were to use it in the old geom library, this would mean that anything using it would be using two different implementations of `Point`/`Vector`. Basically, I think that if we can cut over all the usages of `Point` and `Vector` to the new implementations, we can integrate this class into the old geom stuff immediately. @jsn0716 @scveloso @MathewMacDougall - your thoughts please?

This class doesn't actually use `Point` or `Vector`

**Comment by MathewMacDougall** [comment]
> @jonathanlew - good points. The idea here was that this was going to be integrated into the new geom hierarchy, which means that it must depend on the new Point/Vector implementations. With that in mind, if we were to use it in the old geom library, this would mean that anything using it would be using two different implementations of Point/Vector. Basically, I think that if we can cut over all the usages of Point and Vector to the new implementations, we can integrate this class into the old geom stuff immediately. @jsn0716 @scveloso @MathewMacDougall - your thoughts please?

You're right that if we cut over the new `Point` and `Vector` then that should be fine. We should think ahead if this will cause issues with later geom changes though. Unless we start modifying all geometry "in place" without staging in `new_geom`, other parts of the `new_geom` class hierarchy could get mixed between old and new geom instead.

I think it would be safe to do that, since the `ray`, `polynomial -> line`, and `Spline -> LineSegment` structures are all pretty small and self contained, and shouldn't need too many changes in the rest of the code to change in place.

Why not move the `Splins` to `new_geom` since they'll need to be modified like Polynomial to fit the new structure? Then new navigator stuff can just depend on `new_geom`?

IMO the biggest risk is when all the geom_util stuff gets ported, and how that will go between new and old geom.

**Comment by garethellis0** [comment]
I'd prefer to do one of two things here:
1) integrate `new_geom` stuff fully as we create it
2) construct the entire geom architecture in `new_geom` then integrate it. 

I'm afraid that doing anything in-between these two options will cause more pain then it's worth. Given that we need this immediately (for the navigator stuff), I'm leaning towards (1) currently.

**Comment by MathewMacDougall** [comment]
@jsn0716 @scveloso thoughts on this?

**Comment by jsn0716** [comment]
I think in that case, integrating as we go might be a good idea like @garethellis0 mentioned. It's essentially the same work anyways but in a different order.

**Comment by garethellis0** [comment]
OK, let's merge this separately and then do integration? @jonathanlew - does that work for you? @jsn0716 @scveloso - can you guys handle integrating Point/Vector?

**Comment by jonathanlew** [comment]
> OK, let's merge this separately and then do integration? @jonathanlew - does that work for you? @jsn0716 @scveloso - can you guys handle integrating Point/Vector?

Leave in new_geom and remove the constraint1 constraint2 constructor. Then, make a spline constructor that interpolates points with given degree polynomials and determines coefficients, and that will use the coefficient constructor. That spline constructor can then be easily integrated into path planner

**Comment by JasonCPEN221** [comment]
Sounds good 👍

**Comment by jsn0716** [comment]
Sounds good 👍

**Comment by MathewMacDougall** [comment]
Is that going to be done in this PR or a separate one?

**Comment by jonathanlew** [comment]
> Is that going to be done in this PR or a separate one?

I expect the constraint1 constraint2 constructor to be removed in this PR and then in another PR (for spline), add a spline constructor that does interpolation.

File: src/software/new_geom/BUILD
Comment (jsn0716) [reply src/software/new_geom/BUILD:18]: Removing it caused no errors, removed in: 8c415dfe22b7c11d92acd4f9cfd53990d5ac821f

File: src/software/new_geom/polynomial.h
Comment (jsn0716) [reply src/software/new_geom/polynomial.h]: Removed in af20cdc76f49a87c92abe851d79b90511c7bfb33

File: src/software/new_geom/polynomial.h
Comment (jsn0716) [reply src/software/new_geom/polynomial.h]: Done in af20cdc76f49a87c92abe851d79b90511c7bfb33, although I believe it should be `index` not `index-1`

File: src/software/new_geom/polynomial.cpp
Comment (jsn0716) [reply src/software/new_geom/polynomial.cpp]: Done in af20cdc76f49a87c92abe851d79b90511c7bfb33, removed ternary operator

**Comment by scveloso** [comment]
> OK, let's merge this separately and then do integration? @jonathanlew - does that work for you? @jsn0716 @scveloso - can you guys handle integrating Point/Vector?

@garethellis0 Sorry the mentions got lost in my inbox. Talked to Jason about it - I'll handle replacing old Point/Vector with new Point/Vector.

File: src/software/new_geom/polynomial.h
Comment (garethellis0) [reply src/software/new_geom/polynomial.h]: Doesn't look like this was changed?

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada: (pending other comments here)

File: src/software/new_geom/polynomial.h
Comment (JasonCPEN221) [reply src/software/new_geom/polynomial.h]: My mistake, edited the wrong comment. Fixed here: 1a85e1b1017c773acfad28be47a5f425a06afa67

Comment (jsn0716) [reply src/software/new_geom/polynomial.h]: My mistake, edited the wrong comment. Fixed here: 1a85e1b

**Comment by jsn0716** [comment]
Will need to do one more round of resolving merge conflicts after #986

**Review by jonathanlew — APPROVED** [review]

**Review by garethellis0 — APPROVED** [review]

**Pull request merged**
